pkg/virt-controller/watch/clone/clone.go

```diff
@@ -0,0 +1,365 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package clone
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/utils/pointer"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clonev1alpha1 "kubevirt.io/api/clone/v1alpha1"
+	k6sv1 "kubevirt.io/api/core/v1"
+	snapshotv1alpha1 "kubevirt.io/api/snapshot/v1alpha1"
+	"kubevirt.io/client-go/log"
+)
+
+type cloneSourceType string
+
+const (
+	sourceTypeVM cloneSourceType = "VirtualMachine"
+)
+
+type cloneTargetType string
+
+const (
+	targetTypeVM cloneTargetType = "VirtualMachine"
+	defaultType  cloneTargetType = targetTypeVM
+)
+
+type syncInfo struct {
+	err             error
+	snapshotName    string
+	snapshotReady   bool
+	restoreName     string
+	restoreReady    bool
+	targetVMName    string
+	targetVMCreated bool
+
+	isCloneFailing bool
+	failEvent      Event
+	failReason     string
+
+	reenqueueInfo reenqueueInfo
+}
+type reenqueueInfo struct {
+	reenqueueReason string
+}
+
+func (ctrl *VMCloneController) execute(key string) (error, reenqueueInfo) {
+	var syncInfo syncInfo
+	doNotReenqueue := reenqueueInfo{}
+	logger := log.Log
+
+	obj, cloneExists, err := ctrl.vmCloneInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return err, doNotReenqueue
+	}
+
+	var vmClone *clonev1alpha1.VirtualMachineClone
+	if cloneExists {
+		vmClone = obj.(*clonev1alpha1.VirtualMachineClone)
+		logger = logger.Object(vmClone)
+	} else {
+		return nil, doNotReenqueue
+	}
+
+	var syncErr error
+
+	sourceInfo := vmClone.Spec.Source
+	switch cloneSourceType(sourceInfo.Kind) {
+	case sourceTypeVM:
+		vmKey := getKey(sourceInfo.Name, vmClone.Namespace)
+		obj, vmExists, err := ctrl.vmInformer.GetStore().GetByKey(vmKey)
+		if err != nil {
+			return fmt.Errorf("error getting VM %s in namespace %s from cache: %v", sourceInfo.Name, vmClone.Namespace, err), doNotReenqueue
+		}
+		if !vmExists {
+			return fmt.Errorf("VM %s in namespace %s does not exist", sourceInfo.Name, vmClone.Namespace), doNotReenqueue
```

Comment:
Looks like the webhook allows cloning from a source that doesn't exist.  If we want to support that, may want to watch VMs.  Or requeue.  Otherwise with exponential backoff it may take awhie for the clone to kick in once the source is created

Reply:
Nice catch.
I think it's better be checked by validating webhooks. For example, when creating a migration object, the source VM is expected to exist. We can allow this in the future if we need to.

In any case, I think that there should be a logic that reacts to the scenario where the source VM doesn't exist just to avoid relying on the assumption that it must exist when clone is created. Maybe the proper behavior in such a case is to immediately fail the clone. Does it make sense?

Reply:
I think rejecting creation in the webhook is good for sources that do not exist.  Of course, there are always race conditions between the webhook and the controller, perhaps set progressing false with error that the source does not exist.